File: coding-interview-stack-queue-plus/src/algorithms/stack-queue-problems.js
```javascript
const Stack = require('../data-structures/stack');
const Queue = require('../data-structures/queue');
const Deque = require('../data-structures/deque'); // For Sliding Window Maximum

// =====================================================================================
// Problem 1: Implement Queue using Stacks (LeetCode 232)
// =====================================================================================

/**
 * Implements a FIFO (First-In, First-Out) queue using two LIFO (Last-In, First-Out) stacks.
 * This approach ensures that the oldest element is always at the top of the `outStack`
 * when a dequeue or peek operation is requested.
 */
class MyQueue {
    constructor() {
        this.inStack = new Stack();  // Used for pushing new elements
        this.outStack = new Stack(); // Used for popping/peeking elements
    }

    /**
     * Pushes element x to the back of the queue.
     * This operation is straightforward: simply add to the `inStack`.
     * Time Complexity: O(1) amortized. (Stack push is O(1)).
     * Space Complexity: O(1) for this operation.
     * @param {number} x The element to push.
     */
    push(x) {
        this.inStack.push(x);
    }

    /**
     * Removes the element from the front of the queue and returns it.
     * If `outStack` is empty, all elements from `inStack` are moved to `outStack`
     * to reverse their order, making the oldest element accessible.
     * Time Complexity: O(1) amortized. In the worst case (when `outStack` is empty),
     *                  moving N elements from `inStack` to `outStack` takes O(N).
     *                  However, each element is moved at most once, leading to
     *                  amortized O(1) per pop operation over a sequence of operations.
     * Space Complexity: O(1) for this operation.
     * @returns {number} The element removed from the front.
     */
    pop() {
        this._transferElements(); // Ensure outStack is ready
        return this.outStack.pop();
    }

    /**
     * Returns the element at the front of the queue without removing it.
     * Similar logic to `pop()`, ensures `outStack` has elements.
     * Time Complexity: O(1) amortized. (Same reasoning as `pop`).
     * Space Complexity: O(1) for this operation.
     * @returns {number} The element at the front.
     */
    peek() {
        this._transferElements(); // Ensure outStack is ready
        return this.outStack.peek();
    }

    /**
     * Returns `true` if the queue is empty, `false` otherwise.
     * The queue is empty if both internal stacks are empty.
     * Time Complexity: O(1).
     * Space Complexity: O(1).
     * @returns {boolean} `true` if empty, `false` otherwise.
     */
    empty() {
        return this.inStack.isEmpty() && this.outStack.isEmpty();
    }

    /**
     * Helper function to transfer elements from `inStack` to `outStack`.
     * This is done only when `outStack` is empty, ensuring that elements
     * are reversed only when necessary.
     * All elements from `inStack` are popped and pushed onto `outStack`.
     * This effectively reverses their order, so the first element pushed into `inStack`
     * becomes the last one pushed into `outStack`, thus appearing at its top.
     * Time Complexity: O(N) where N is the number of elements in `inStack`.
     * Space Complexity: O(1) beyond the stacks themselves.
     * @private
     */
    _transferElements() {
        if (this.outStack.isEmpty()) {
            while (!this.inStack.isEmpty()) {
                this.outStack.push(this.inStack.pop());
            }
        }
    }
}


// =====================================================================================
// Problem 2: Implement Stack using Queues (LeetCode 225)
// =====================================================================================

/**
 * Implements a LIFO (Last-In, First-Out) stack using two FIFO (First-In, First-Out) queues.
 * This approach ensures that the most recently added element is always at the front of `q1`.
 *
 * Approach chosen: Push O(N), Pop O(1).
 * When pushing an element `x`:
 * 1. Enqueue `x` to `q1`.
 * 2. Move all existing elements from `q1` (that were added before `x`) to `q2`.
 *    This makes `x` the only element in `q1`.
 * 3. Move all elements from `q2` back to `q1`.
 *    This effectively puts `x` at the front of `q1`, followed by all previously added elements
 *    in their original relative order.
 */
class MyStack {
    constructor() {
        this.q1 = new Queue(); // Main queue, stores elements in stack order (newest at front)
        this.q2 = new Queue(); // Auxiliary queue, used for reordering during push
    }

    /**
     * Pushes element x onto the stack.
     * Time Complexity: O(N) where N is the current number of elements in the stack.
     *                  Each push operation involves moving all existing N elements twice.
     * Space Complexity: O(1) for this operation (ignoring the space for N elements).
     * @param {number} x The element to push.
     */
    push(x) {
        // Step 1: Add new element to q1
        this.q1.enqueue(x);

        // Step 2 & 3: Move all elements except the newly added one from q1 to q2, then back to q1.
        // This ensures the new element 'x' is at the front of q1.
        // We only move (size - 1) elements because 'x' is already at the end of q1.
        // After 'x' is enqueued, q1.size() will be old_size + 1.
        // We need to move 'old_size' elements to q2 and then back.
        // The element 'x' itself should remain in q1 to be at the front.
        // So we iterate `old_size` times.
        let currentSize = this.q1.size();
        while (currentSize > 1) { // Leave the 'x' at the end of q1
            this.q2.enqueue(this.q1.dequeue());
            currentSize--;
        }

        // Now 'x' is the only element left in q1, and all others are in q2.
        // Transfer elements from q2 back to q1.
        while (!this.q2.isEmpty()) {
            this.q1.enqueue(this.q2.dequeue());
        }

        // After these operations, q1 contains [x, old_top, ..., old_bottom]
        // where x is the new top, and old_top is the previous top.
    }

    /**
     * Removes the element on top of the stack and returns it.
     * Since `q1` is maintained such that the top element is always at its front,
     * this is a simple dequeue operation.
     * Time Complexity: O(1).
     * Space Complexity: O(1).
     * @returns {number} The element removed from the top.
     */
    pop() {
        if (this.empty()) {
            return undefined; // Stack is empty
        }
        return this.q1.dequeue();
    }

    /**
     * Returns the element on top of the stack without removing it.
     * Similarly, this is a simple peek operation on `q1`.
     * Time Complexity: O(1).
     * Space Complexity: O(1).
     * @returns {number} The element at the top.
     */
    top() {
        if (this.empty()) {
            return undefined; // Stack is empty
        }
        return this.q1.peek();
    }

    /**
     * Returns `true` if the stack is empty, `false` otherwise.
     * The stack is empty if `q1` is empty (since all elements are kept in `q1`).
     * Time Complexity: O(1).
     * Space Complexity: O(1).
     * @returns {boolean} `true` if empty, `false` otherwise.
     */
    empty() {
        return this.q1.isEmpty();
    }

    /**
     * Alternative Approach: Push O(1), Pop O(N).
     *
     * In this approach:
     * - `push(x)`: Enqueue `x` to `q1`. O(1).
     * - `pop()`: To get the top element, we must dequeue all but the last element
     *            from `q1` and enqueue them into `q2`. The last remaining element
     *            in `q1` is the top. Dequeue it. Then swap `q1` and `q2`. O(N).
     * - `top()`: Same as `pop` but re-enqueue the last element back to `q1` or `q2`
     *            before swapping. O(N).
     *
     * The chosen approach (Push O(N), Pop O(1)) is often preferred in interviews
     * because pop/top are frequently called, and O(1) is highly desirable for these operations.
     */
}


// =====================================================================================
// Problem 3: Valid Parentheses (LeetCode 20)
// =====================================================================================

/**
 * Given a string `s` containing just the characters '(', ')', '{', '}', '[' and ']',
 * determine if the input string is valid.
 * An input string is valid if:
 * 1. Open brackets must be closed by the same type of brackets.
 * 2. Open brackets must be closed in the correct order.
 * 3. Every close bracket has a corresponding open bracket of the same type.
 *
 * This is a classic stack problem.
 *
 * Optimal Solution:
 * Iterate through the string.
 * If an opening bracket is encountered, push it onto the stack.
 * If a closing bracket is encountered:
 *   - If the stack is empty, it's an invalid string (no matching opening bracket).
 *   - Pop the top element from the stack.
 *   - Check if the popped opening bracket matches the current closing bracket.
 *     If not, it's an invalid string.
 * After iterating through the entire string, if the stack is empty, the string is valid.
 * Otherwise, there are unclosed opening brackets, so it's invalid.
 *
 * @param {string} s The input string.
 * @returns {boolean} True if the string is valid, false otherwise.
 *
 * Time Complexity: O(N) where N is the length of the string `s`.
 *                  Each character is processed once, with O(1) stack operations.
 * Space Complexity: O(N) in the worst case. E.g., for "((((((", the stack will
 *                   store all N opening brackets. In the best case (e.g., "()()"),
 *                   space is O(1) if elements are pushed and immediately popped.
 */
function isValid(s) {
    const stack = new Stack();
    // Map opening brackets to their corresponding closing brackets.
    const bracketMap = {
        '(': ')',
        '{': '}',
        '[': ']'
    };

    // Iterate through each character in the string
    for (let i = 0; i < s.length; i++) {
        const char = s[i];

        // If the character is an opening bracket, push it onto the stack
        if (bracketMap[char]) {
            stack.push(char);
        } else {
            // If the character is a closing bracket
            // Check if the stack is empty. If so, no opening bracket to match.
            if (stack.isEmpty()) {
                return false;
            }

            // Pop the top element from the stack. This should be the corresponding opening bracket.
            const lastOpenBracket = stack.pop();

            // Check if the popped opening bracket matches the current closing bracket.
            // E.g., if char is ')' and lastOpenBracket is '(', they match.
            // If bracketMap['('] (which is ')') != char, then it's a mismatch.
            if (bracketMap[lastOpenBracket] !== char) {
                return false;
            }
        }
    }

    // After iterating through the entire string, if the stack is empty,
    // all opening brackets have been correctly closed.
    return stack.isEmpty();
}


// =====================================================================================
// Problem 4: Sliding Window Maximum (LeetCode 239)
// =====================================================================================

/**
 * You are given an array of integers `nums`, there is a sliding window of size `k`
 * which is moving from the very left of the array to the very right.
 * Return the max sliding window.
 *
 * Example:
 * nums = [1,3,-1,-3,5,3,6,7], k = 3
 * Window position          Max
 * ---------------          -----
 * [1  3  -1] -3  5  3  6  7   3
 *  1 [3  -1  -3] 5  3  6  7   3
 *  1  3 [-1  -3  5] 3  6  7   5
 *  1  3  -1 [-3  5  3] 6  7   5
 *  1  3  -1  -3 [5  3  6] 7   6
 *  1  3  -1  -3  5 [3  6  7]  7
 *
 * Output: [3,3,5,5,6,7]
 *
 * Optimal Solution: Using a Monotonic Deque (Double-Ended Queue).
 * A deque is used to store indices of elements. The deque maintains elements in
 * decreasing order from front to back.
 *
 * Algorithm:
 * 1. Initialize an empty `Deque` to store indices and an empty `result` array.
 * 2. Iterate through `nums` with index `i` from `0` to `n-1`:
 *    a. **Remove elements outside the window:** If the deque is not empty and its
 *       front element's index (`deque.peekFront()`) is `i - k` (meaning it's
 *       no longer in the current window), remove it from the front (`deque.removeFront()`).
 *    b. **Maintain monotonic decreasing order:** While the deque is not empty AND
 *       the element at the back of the deque (`nums[deque.peekBack()]`) is less than
 *       or equal to the current element (`nums[i]`), remove elements from the back
 *       (`deque.removeBack()`). This ensures the deque only contains relevant
 *       (potentially maximum) elements, and the largest is at the front.
 *    c. **Add current element:** Add the current index `i` to the back of the deque (`deque.addBack(i)`).
 *    d. **Record maximum:** If `i` is greater than or equal to `k - 1` (i.e., the window has
 *       fully formed and is moving), the maximum element in the current window is
 *       `nums[deque.peekFront()]`. Add this to the `result` array.
 * 3. Return the `result` array.
 *
 * @param {number[]} nums The input array of integers.
 * @param {number} k The size of the sliding window.
 * @returns {number[]} An array containing the maximum for each sliding window.
 *
 * Time Complexity: O(N) where N is the length of `nums`.
 *                  Each element is added to and removed from the deque at most once.
 * Space Complexity: O(K) in the worst case, as the deque can store up to `k` elements/indices.
 */
function maxSlidingWindow(nums, k) {
    if (nums === null || nums.length === 0 || k <= 0) {
        return [];
    }
    if (k === 1) {
        return nums; // Each element is its own max
    }

    const result = [];
    const deque = new Deque(); // Stores indices of elements

    for (let i = 0; i < nums.length; i++) {
        // 1. Remove elements from the front if they are out of the current window
        // The element at deque.peekFront() is the index of the current maximum.
        // If this index is less than i - k + 1 (or equivalently, deque.peekFront() <= i - k),
        // it means it's outside the window [i-k+1, i].
        if (!deque.isEmpty() && deque.peekFront() <= i - k) {
            deque.removeFront();
        }

        // 2. Remove elements from the back that are smaller than the current number nums[i].
        // These elements can never be the maximum in any future window that includes nums[i]
        // because nums[i] is greater and appears later.
        while (!deque.isEmpty() && nums[deque.peekBack()] <= nums[i]) {
            deque.removeBack();
        }

        // 3. Add the current element's index to the back of the deque.
        deque.addBack(i);

        // 4. If the window has fully formed (i.e., `i` has reached `k-1` or beyond),
        // the maximum for the current window is at the front of the deque.
        if (i >= k - 1) {
            result.push(nums[deque.peekFront()]);
        }
    }

    return result;
}

// =====================================================================================
// Problem 5: Trapping Rain Water (LeetCode 42)
// =====================================================================================

/**
 * Given `n` non-negative integers representing an elevation map where the width of each bar is `1`,
 * compute how much water it can trap after raining.
 *
 * Example: height = [0,1,0,2,1,0,1,3,2,1,2,1]
 * Output: 6
 *
 * Visualization:
 *          _
 *      _  | | _
 *    _|=| | | | | _
 * _|_|S|_|_|S|S|_|_|_
 * [0,1,0,2,1,0,1,3,2,1,2,1]
 * S = water trapped
 *
 * Optimal Solution: Using a Monotonic Stack.
 * The idea is to find potential "wells" (valleys) and calculate the water they can hold.
 * A monotonic decreasing stack stores indices of bars.
 *
 * Algorithm:
 * 1. Initialize `water = 0` and an empty `Stack`.
 * 2. Iterate through the `height` array with index `i` from `0` to `n-1`:
 *    a. **While the stack is NOT empty AND `height[i]` is GREATER THAN `height[stack.peek()]`:**
 *       This means we found a right boundary (`height[i]`) that is higher than the bar
 *       at the top of the stack, forming a potential well.
 *       i. Pop `prevIdx` from the stack. This `prevIdx` represents the bottom of the well.
 *       ii. If the stack becomes empty after popping `prevIdx`, it means there's no left
 *           boundary for this well (or the left boundary is further left than any previous
 *           bar, effectively making it unbounded to the left for this calculation). Break this inner loop.
 *       iii. Otherwise, `leftIdx = stack.peek()`. This is our left boundary.
 *       iv. Calculate the distance (width) of the well: `distance = i - leftIdx - 1`.
 *           (Current index `i` is right boundary, `leftIdx` is left boundary, minus 1 for the `prevIdx` bar itself).
 *       v. Calculate the height of the water above `prevIdx`: `h = Math.min(height[i], height[leftIdx]) - height[prevIdx]`.
 *          The water level is limited by the shorter of the two boundaries.
 *       vi. Add trapped water: `water += distance * h`.
 *    b. **Push current index `i` onto the stack.** This maintains the monotonic decreasing property.
 * 3. Return `water`.
 *
 * @param {number[]} height An array of non-negative integers representing the elevation map.
 * @returns {number} The total amount of water trapped.
 *
 * Time Complexity: O(N) where N is the length of `height`.
 *                  Each element is pushed onto and popped from the stack at most once.
 * Space Complexity: O(N) in the worst case, if the heights are strictly decreasing
 *                   (e.g., [5,4,3,2,1]), the stack will store all N elements.
 */
function trap(height) {
    let water = 0;
    const stack = new Stack(); // Stores indices of bars in decreasing order of height

    for (let i = 0; i < height.length; i++) {
        // While the stack is not empty AND the current bar is taller than the bar at the stack's top
        // This indicates we've found a right boundary for a potential well.
        while (!stack.isEmpty() && height[i] > height[stack.peek()]) {
            const prevIdx = stack.pop(); // This is the bottom of our current well.

            // If the stack becomes empty, it means there's no left boundary for this `prevIdx`.
            // So, no water can be trapped with `prevIdx` as the bottom.
            if (stack.isEmpty()) {
                break;
            }

            const leftIdx = stack.peek(); // The bar to the left acting as a boundary.

            // Calculate the width of the well
            // (Current right boundary index - left boundary index - 1 for the `prevIdx` bar itself)
            const distance = i - leftIdx - 1;

            // Calculate the height of the water that can be trapped above `prevIdx` bar
            // It's limited by the shorter of the two boundaries (left and right),
            // minus the height of the `prevIdx` bar itself.
            const trappedHeight = Math.min(height[i], height[leftIdx]) - height[prevIdx];

            // Add the water trapped in this segment
            water += distance * trappedHeight;
        }

        // Push the current bar's index onto the stack.
        // This maintains the monotonic decreasing order (or stores candidates for left boundaries).
        stack.push(i);
    }

    return water;
}


module.exports = {
    MyQueue,
    MyStack,
    isValid,
    maxSlidingWindow,
    trap,
};
```

File: leetcode-stack-queue-solutions-plus/src/stack_queue_problems.py
```python
"""
stack_queue_problems.py

This file contains optimal solutions to several common coding interview problems
involving Stack and Queue data structures. Each function includes detailed
comments, time, and space complexity analysis.
"""

import collections

# --- Problem 1: Valid Parentheses ---
# LeetCode: 20
# Description: Given a string s containing just the characters '(', ')', '{', '}', '[' and ']',
# determine if the input string is valid.
# An input string is valid if:
# 1. Open brackets must be closed by the same type of brackets.
# 2. Open brackets must be closed in the correct order.
# 3. Every close bracket has a corresponding open bracket of the same type.

def is_valid_parentheses(s: str) -> bool:
    """
    Determines if the input string has valid parentheses.

    Uses a stack to keep track of open brackets. When a closing bracket is encountered,
    it checks if the stack's top element is the corresponding opening bracket.

    Args:
        s: The input string containing parentheses.

    Returns:
        True if the parentheses are valid, False otherwise.

    Time Complexity: O(N), where N is the length of the string `s`.
                     Each character is pushed onto the stack or popped from it once.
    Space Complexity: O(N), in the worst case (e.g., "((("), the stack can store
                      all opening brackets.
    """
    stack = []  # Use a list as a stack
    mapping = {")": "(", "}": "{", "]": "["} # Map closing brackets to their opening counterparts

    for char in s:
        if char in mapping:  # It's a closing bracket
            # Get the top element from the stack, or use a dummy value if stack is empty
            top_element = stack.pop() if stack else '#'
            # Check if the popped element matches the expected opening bracket
            if mapping[char] != top_element:
                return False
        else:  # It's an opening bracket
            stack.append(char)

    # After iterating through the entire string, the stack should be empty
    # if all opening brackets were correctly closed.
    return not stack

# --- Problem 2: Min Stack ---
# LeetCode: 155
# Description: Design a stack that supports push, pop, top, and retrieving the minimum
# element in constant time.
# Implement the MinStack class:
# - MinStack() initializes the stack object.
# - void push(int val) pushes the element val onto the stack.
# - void pop() removes the element on the top of the stack.
# - int top() gets the top element of the stack.
# - int getMin() retrieves the minimum element in the stack.

class MinStack:
    """
    Implements a stack with O(1) time complexity for push, pop, top, and getMin.

    Approach: Use an auxiliary stack (`min_stack`) to keep track of the minimum
    element at each corresponding state of the main stack.

    Time Complexity:
        - push: O(1)
        - pop: O(1)
        - top: O(1)
        - getMin: O(1)
    Space Complexity: O(N), where N is the number of elements in the stack.
                      In the worst case (elements pushed in descending order),
                      both stacks can grow to size N.
    """
    def __init__(self):
        """
        Initializes the MinStack object with an empty main stack and an empty
        auxiliary minimum tracking stack.
        """
        self.stack = []       # Main stack to store all elements
        self.min_stack = []   # Auxiliary stack to store the minimum at each point

    def push(self, val: int) -> None:
        """
        Pushes an element onto the stack. Updates the min_stack accordingly.
        """
        self.stack.append(val)
        # The new minimum is the smaller of the current value and the current min_stack top
        # (if min_stack is not empty). If min_stack is empty, the current value is the min.
        current_min = self.min_stack[-1] if self.min_stack else float('inf')
        self.min_stack.append(min(val, current_min))

    def pop(self) -> None:
        """
        Removes the element from the top of the stack.
        Both stacks are popped to maintain consistency.
        Raises an IndexError if the stack is empty.
        """
        if not self.stack:
            raise IndexError("pop from empty stack")
        self.stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        """
        Returns the element at the top of the stack without removing it.
        Raises an IndexError if the stack is empty.
        """
        if not self.stack:
            raise IndexError("top from empty stack")
        return self.stack[-1]

    def getMin(self) -> int:
        """
        Retrieves the current minimum element in the stack.
        Raises an IndexError if the stack is empty.
        """
        if not self.min_stack:
            raise IndexError("getMin from empty stack")
        return self.min_stack[-1]

# --- Problem 3: Implement Queue using Stacks ---
# LeetCode: 232
# Description: Implement a first-in-first-out (FIFO) queue using only two stacks.
# The implemented queue should support all the functions of a normal queue
# (push, peek, pop, and empty).
#
# Implement the MyQueue class:
# - push(x) Pushes element x to the back of the queue.
# - pop() Removes the element from the front of the queue and returns it.
# - peek() Returns the element at the front of the queue.
# - empty() Returns true if the queue is empty, false otherwise.
#
# Notes:
# - You must use only standard operations of a stack, which means only push to top,
#   peek/pop from top, size, and is_empty operations are valid.
# - You may assume that all operations are valid (for example, no pop or peek operations
#   will be called on an empty queue).

class MyQueue:
    """
    Implements a Queue using two stacks. This approach provides amortized O(1)
    time complexity for push, pop, and peek operations.

    Approach:
    - `in_stack`: Used for push operations (enqueue elements).
    - `out_stack`: Used for pop and peek operations (dequeue elements).

    When popping or peeking, if `out_stack` is empty, all elements are moved
    from `in_stack` to `out_stack`. This reversal ensures FIFO order.
    The cost of moving N elements is amortized over N push operations.

    Time Complexity:
        - push: O(1)
        - pop: Amortized O(1). Worst case O(N) when `out_stack` is empty and
               elements need to be moved from `in_stack`.
        - peek: Amortized O(1). Worst case O(N).
        - empty: O(1)
    Space Complexity: O(N), where N is the number of elements in the queue.
                      Both stacks together store all elements.
    """
    def __init__(self):
        """
        Initializes the queue with two empty stacks.
        """
        self.in_stack = []   # For new elements (push)
        self.out_stack = []  # For old elements (pop, peek)

    def push(self, x: int) -> None:
        """
        Pushes element x to the back of the queue.
        Time: O(1)
        """
        self.in_stack.append(x)

    def _transfer_if_needed(self) -> None:
        """
        Helper method to transfer elements from in_stack to out_stack if out_stack is empty.
        This ensures elements are in the correct order for FIFO operations.
        """
        if not self.out_stack:
            while self.in_stack:
                self.out_stack.append(self.in_stack.pop())

    def pop(self) -> int:
        """
        Removes the element from the front of the queue and returns it.
        Time: Amortized O(1) (Worst case O(N))
        """
        self._transfer_if_needed()
        if not self.out_stack:
            raise IndexError("pop from empty queue") # Should not happen per problem statement
        return self.out_stack.pop()

    def peek(self) -> int:
        """
        Returns the element at the front of the queue.
        Time: Amortized O(1) (Worst case O(N))
        """
        self._transfer_if_needed()
        if not self.out_stack:
            raise IndexError("peek from empty queue") # Should not happen per problem statement
        return self.out_stack[-1]

    def empty(self) -> bool:
        """
        Returns true if the queue is empty, false otherwise.
        Time: O(1)
        """
        return not self.in_stack and not self.out_stack


# --- Problem 4: Sliding Window Maximum ---
# LeetCode: 239
# Description: You are given an array of integers nums, there is a sliding window of
# size k which is moving from the very left of the array to the very right.
# You can only see the k numbers in the window. Each time the sliding window moves
# right by one position.
# Return the max sliding window.

def max_sliding_window(nums: list[int], k: int) -> list[int]:
    """
    Finds the maximum value in each sliding window of size k.

    Uses a deque (double-ended queue) to store indices of elements in the current
    window in decreasing order of their values. This is a Monotonic Deque approach.

    Args:
        nums: The input list of integers.
        k: The size of the sliding window.

    Returns:
        A list of maximum values for each sliding window.

    Time Complexity: O(N), where N is the length of `nums`. Each element is
                     added and removed from the deque at most once.
    Space Complexity: O(K), as the deque stores at most K elements (indices).
    """
    if not nums or k == 0:
        return []

    if k == 1: # Edge case: window size 1, just return the array
        return nums

    # Deque stores indices, not values. Elements in deque are in decreasing order.
    # The front of the deque (index 0) will always hold the index of the maximum element
    # in the current window.
    dq = collections.deque()
    result = []
    n = len(nums)

    for i in range(n):
        # 1. Remove elements from the front of the deque if they are outside the current window.
        #    The current window ranges from (i - k + 1) to i.
        if dq and dq[0] == i - k:
            dq.popleft()

        # 2. Remove elements from the back of the deque that are smaller than the current element `nums[i]`.
        #    These elements can no longer be the maximum because `nums[i]` is greater and more recent.
        while dq and nums[dq[-1]] <= nums[i]:
            dq.pop()

        # 3. Add the current element's index to the back of the deque.
        dq.append(i)

        # 4. If the window has fully formed (i.e., we have processed at least `k` elements),
        #    the maximum for this window is `nums[dq[0]]`.
        if i >= k - 1:
            result.append(nums[dq[0]])

    return result

# --- Problem 5: Daily Temperatures ---
# LeetCode: 739
# Description: Given an array of integers temperatures represents the daily temperatures,
# return an array answer such that answer[i] is the number of days you have to wait
# after the ith day to get a warmer temperature. If there is no future day for which
# this is possible, keep answer[i] == 0 instead.

def daily_temperatures(temperatures: list[int]) -> list[int]:
    """
    Calculates the number of days to wait for a warmer temperature.

    Uses a monotonic stack (specifically, a decreasing stack) to efficiently find
    the next greater element for each temperature. The stack stores indices.

    Args:
        temperatures: A list of daily temperatures.

    Returns:
        A list where `answer[i]` is the number of days to wait for a warmer
        temperature after day `i`.

    Time Complexity: O(N), where N is the number of temperatures. Each temperature
                     is pushed onto the stack and popped from it at most once.
    Space Complexity: O(N), in the worst case (e.g., temperatures in decreasing
                      order), the stack can store all indices.
    """
    n = len(temperatures)
    answer = [0] * n  # Initialize result array with zeros
    stack = []        # Stores indices of temperatures in decreasing order

    for i in range(n):
        current_temp = temperatures[i]

        # While stack is not empty and the current temperature is warmer than
        # the temperature at the index on top of the stack:
        while stack and current_temp > temperatures[stack[-1]]:
            # Pop the index from the stack
            prev_index = stack.pop()
            # Calculate the waiting days and store it in the answer array
            answer[prev_index] = i - prev_index

        # Push the current index onto the stack
        stack.append(i)

    # After the loop, any indices remaining in the stack have no warmer day
    # to their right, so their corresponding answer[i] remains 0 (as initialized).
    return answer

if __name__ == "__main__":
    from src.utils import Timer

    print("--- Problem 1: Valid Parentheses ---")
    with Timer("Valid Parentheses"):
        test_cases_vp = [
            ("()", True),
            ("()[]{}", True),
            ("(]", False),
            ("([)]", False),
            ("{[]}", True),
            ("", True),
            ("[", False),
            ("]", False),
            ("((((", False),
            ("))))", False),
            ("{[()]}", True)
        ]
        for s, expected in test_cases_vp:
            result = is_valid_parentheses(s)
            print(f"'{s}' -> {result} (Expected: {expected}) {'✅' if result == expected else '❌'}")

    print("\n--- Problem 2: Min Stack ---")
    with Timer("Min Stack"):
        minStack = MinStack()
        minStack.push(-2)
        minStack.push(0)
        minStack.push(-3)
        print(f"Min: {minStack.getMin()}") # Expected: -3
        minStack.pop()
        print(f"Top: {minStack.top()}")   # Expected: 0
        print(f"Min: {minStack.getMin()}") # Expected: -2
        minStack.pop()
        print(f"Top: {minStack.top()}")   # Expected: -2
        print(f"Min: {minStack.getMin()}") # Expected: -2
        minStack.push(-1)
        minStack.push(10)
        print(f"Top: {minStack.top()}")   # Expected: 10
        print(f"Min: {minStack.getMin()}") # Expected: -2

        minStack2 = MinStack()
        minStack2.push(5)
        minStack2.push(4)
        minStack2.push(3)
        minStack2.push(2)
        minStack2.push(1)
        print(f"Min2: {minStack2.getMin()}") # Expected: 1
        minStack2.pop() # pop 1
        print(f"Min2: {minStack2.getMin()}") # Expected: 2
        minStack2.pop() # pop 2
        print(f"Min2: {minStack2.getMin()}") # Expected: 3


    print("\n--- Problem 3: Implement Queue using Stacks ---")
    with Timer("Queue using Stacks"):
        myQueue = MyQueue()
        myQueue.push(1) # queue is: [1]
        myQueue.push(2) # queue is: [1, 2] (left is front)
        print(f"Peek: {myQueue.peek()}") # Expected: 1
        print(f"Pop: {myQueue.pop()}")   # Expected: 1, queue is: [2]
        print(f"Empty? {myQueue.empty()}") # Expected: False
        myQueue.push(3) # queue is: [2, 3]
        print(f"Peek: {myQueue.peek()}") # Expected: 2
        print(f"Pop: {myQueue.pop()}")   # Expected: 2, queue is: [3]
        print(f"Pop: {myQueue.pop()}")   # Expected: 3, queue is: []
        print(f"Empty? {myQueue.empty()}") # Expected: True

    print("\n--- Problem 4: Sliding Window Maximum ---")
    with Timer("Sliding Window Maximum"):
        test_cases_swm = [
            ([1,3,-1,-3,5,3,6,7], 3, [3,3,5,5,6,7]),
            ([1], 1, [1]),
            ([1, -1], 1, [1, -1]),
            ([9,11], 2, [11]),
            ([4,-2], 2, [4]),
            ([7,2,4], 2, [7,4]),
            ([1,3,1,2,0,5], 3, [3,3,2,5]),
            ([1,2,3,4,5], 3, [3,4,5]),
            ([5,4,3,2,1], 3, [5,4,3]),
            ([], 0, []),
            ([1,2,3], 0, []) # Should be handled as empty list for k=0
        ]
        for nums, k, expected in test_cases_swm:
            result = max_sliding_window(nums, k)
            print(f"Nums: {nums}, k: {k} -> {result} (Expected: {expected}) {'✅' if result == expected else '❌'}")


    print("\n--- Problem 5: Daily Temperatures ---")
    with Timer("Daily Temperatures"):
        test_cases_dt = [
            ([73,74,75,71,69,72,76,73], [1,1,4,2,1,1,0,0]),
            ([30,40,50,60], [1,1,1,0]),
            ([30,60,90], [1,1,0]),
            ([89,62,70,58,47,47,46,76,100,70], [8,1,5,4,3,2,1,1,0,0]),
            ([100,90,80,70,60], [0,0,0,0,0]),
            ([60,70,60,80], [1,2,1,0]),
            ([70,60,50,40], [0,0,0,0]),
            ([], [])
        ]
        for temperatures, expected in test_cases_dt:
            result = daily_temperatures(temperatures)
            print(f"Temps: {temperatures} -> {result} (Expected: {expected}) {'✅' if result == expected else '❌'}")
```